_posts/2018-10-25-Rust-1.30.md

```diff
@@ -0,0 +1,355 @@
+```
+
+Here, we have two input `TokenStreams`; the first is for the contents of the
+attribute itself, that is, the `GET, "/"` stuff. The second is the body of the
+thing the attribute is attached to, in this case, `fn index() {}` and the rest
+of the function's body.
+
+Function-like macros define macros that look like function calls. For
+example, an `sql!` macro:
+
+```rust
+let sql = sql!(SELECT * FROM posts WHERE id=1);
+```
+
+This macro would parse the SQL statement inside of it and check that it's
+syntactically correct. This macro would be defined like this:
+
+```
+#[proc_macro]
+pub fn sql(input: TokenStream) -> TokenStream {
+```
+
+This is similar to the derive macro's signature: we get in the tokens that
+are inside of the parentheses, and return the code we wanted to generate.
+
+### `use` and macros
+
+You can now [use `use` to bring macros into scope][externmacro]. For example, 
+to use `serde-json`'s `json` macro, you used to write:
+
+```rust
+#[macro_use]
+extern crate serde_json;
+
+let john = json!({
+    "name": "John Doe",
+    "age": 43,
+    "phones": [
+    "+44 1234567",
+    "+44 2345678"
+    ]
+});
+```
+
+But now, you'd write
+
+```rust
+extern crate serde_json;
+
+use serde_json::json;
+
+let john = json!({
+    "name": "John Doe",
+    "age": 43,
+    "phones": [
+    "+44 1234567",
+    "+44 2345678"
+    ]
+});
+```
+
+This brings macros more in line with other items, and removes the need for
+`macro_use` annotations.
+
+[externmacro]: https://github.com/rust-lang/rust/pull/50911/
+[Rust 1.15]: https://blog.rust-lang.org/2017/02/02/Rust-1.15.html
+
+## Module system improvements
+
+The module system has long been a pain point of new Rustaceans; several of
+its rules felt awkward in practice. These changes are the first steps we're
+taking to make the module system feel more straightforward.
+
+[`mod.rs` files are now optional][optionalmod]. Imagine we had a `foo`
+submodule, with a `bar` submodule of its own. The directory layout would
+look like this:
+
+```text
+.
+├── src
+│   ├── foo
+│   │   └── mod.rs
+│   │   └── bar.rs
+│   └── lib.rs
+```
+
+In Rust 1.30, you can do this instead:
+
+```text
+.
+├── src
+│   ├── foo
+│   │   └── bar.rs
+│   ├── foo.rs
+│   └── lib.rs
+```
+
+This means that you'll no longer have plenty of tabs all named `mod.rs` in
+your IDE! It also eases the process of adding a submodule; before `bar`
+existed, the project would look like this:
+
+```text
+.
+├── src
+│   ├── foo.rs
+│   └── lib.rs
+```
+
+Many users found the need to move `foo.rs` into `foo/mod.rs` to be an
+unncessary, strage requirement. With the new layout, you create `src/foo` and
+put `bar.rs` in it and you're done.
+
+[optionalmod]: https://github.com/rust-lang/rust/pull/54072
+
+There's two changes to `use` as well. Well, three changes: we already
+mentioned that you can use `use` to bring macros into scope in the macros
+section. There's two more. The first is that you can [`use` an extern crate
```

Comment:
This not about `use`, this is about anything but `use` actually.
What happens is that extern crate names are put into prelude, so in relative paths they can be used from the whole crate, similarly to e.g. `Vec` or `Copy`.
(Paths in `use` are absolute on 2015 edition, so they are not affected by this change.)